trunk/xchat-gnome/data/xchat-gnome-remote.1
.TH XCHAT-GNOME "1" "April 2006" "xchat-gnome 0.10" "User Commands"
.SH NAME
xchat-gnome-remote \- tells a running xchat-gnome to perform tasks via dbus
.SH DESCRIPTION
.PP
This is a manual page written for Debian because the original package doesn't have one. For further reference, check the \fBSEE ALSO\fR section.
.SS "Usage:"
.IP
xchat\-gnome\-remote [OPTION...]
.SS "Options:"
\fB\-u\fR, \fB\-\-url\fR=\fIirc\fR://server:port/channel
Open an irc:// url
.TP
\fB\-c\fR, \fB\-\-command\fR=\fcommand\fR
Execute a xchat command
.TP
\fB\-p\fR, \fB\-\-print\fR=\ftext\fR
Prints some text to the current tab/window
.TP
\fB\-s\fR, \fB\-\-channel\fR=\fchannel\fR
Change the context to the channel
.TP
\fB\-s\fR, \fB\-\-server\fR=\fserver\fR
Change the context to the server
.TP
\fB\-i\fR, \fB\-\-info\fR=\fid\fR
Get some informations from xchat
.TP
\fB\-r\fR, \fB\-\-prefs\fR=\fname\fR
Get settings from xchat
.TP
.SH "SEE ALSO"
.BR xchat-gnome (1)


NEWS
0.12 "All roads lead to Sean Connery" (20 June 2006)
	* A new sheme for theme-based foreground and backgrond colors, with
	  automatic palette optimization for mIRC color codes.
	* Remove redundant nicknames (with a gconf key to re-enable them)
	* Show a whois tooltip on users in the conversation panel
	* A large number of bug fixes and smaller improvements

0.11 "I have plenty of better things to do, I just aren't doing them" (04 April 2006)
	* Show whois info in a tooltip on the userlist
	* Add a preferences page for controlling spell-checking (enabling it,
	  selecting languages)
	* As usual, numerous bug fixes and minor improvements

0.10 "Almost as great" (24 February 2006)
	* Add a sound notification plugin
	* Session management improvements
	* Minor UI polish
	* A huge slew of bug-fixes

0.9 "You guys sell every kind of meat here except human!" (19 January 2006)
	* Implement channel list filtering
	* Update notify-osd plugin for libnotify >=0.3
	* Add UI for setting nickserv auto-identify password
	* Cooler plugins preferences
	* Lots and lots of bug-fixes and minor improvements

0.8 "Trailblazing where everyone else should follow, dammit" (19 December 2005)
	* New pop-up userlist
	* New, relatively sane build-system
	* Translation support
	* Translations
	* various and sundry bugfixes

0.7 "The best thing I've ever stood on" (20 November 2005)
	* Added a help document
	* Add background preferences
	* Add a plugin which intelligently disconnects/reconnects based on
	  network connection status via NetworkManager
	* Add an "OSD" plugin, which pops up important messages with libnotify
	* Implement "Save Transcript"
	* Add a plugin which sets away status based on screensaver state
	* various and sundry bugfixes

0.6 "Braindome" (23 October 2005)
	* Set up a gnome url handler for irc:// links
	* Add "ignore" to the user context menu
	* Add autoconnect and autojoin options in context menus
	* Add nickname change dialog when clicking on the nick on the left of
	  the input bar
	* Add text and files drag and drop support
	* Beginnings of command completion

0.5 "You are what you eat" (27 August 2005)
	* DCC user interface, including send/recieve window, preferences, file
	  choosers, etc
	* libsexy integration- urls in topics, spell checking
	* lots of bugfixes

0.4 "Buyer beware - Penis patches!" (13 May 2005)
	* Basic functionality in notification area plugin
	* Connect dialog will only connect once to any particular server
	* Closing entire servers in the navigation list
	* Variety of internal reorganization to optimize speed and memory
	* Sort lists in the orders which make sense for them
	* Various UI tweaking
	* lots and lots of bugfixes

0.3 "You be the horse, and I'll be the buggy" (03 March 2005)
	* Plugin autoloading
	* Network configuration
	* Added Firefox-style search bar
	* Conversion to GtkUIManager
	* Hilight list editing
	* Color configuration (schemes, custom)
	* Infrastructure for Keybinding configuration
	* bugfixes

0.2 "This is my boomstick" (13 September 2004)
	* various GUI prettification
	* GUI for selecting character encoding
	* fix build prefix for creating packages
	* implemented various menu items
	* new topic bar
	* new setup dialog
	* font configuration
	* plugin configuration
	* bugfixes

0.1 "Doctors are Standing By" (3 May 2004)
	* first release, generally lots of stuff


plugins/dbus/xchat-gnome-remote.c
/* xchat-remote - program for remote access xchat using DBUS
 * Copyright (C) 2005 Claessens Xavier
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Claessens Xavier
 * xclaesse@gmail.com
 */

#include <config.h>
#include <dbus/dbus-glib.h>
#include <unistd.h>
#include <glib/gi18n.h>

#define DBUS_SERVICE "org.xchatgnome.service"
#define DBUS_OBJECT "/org/xchatgnome/RemoteObject"
#define DBUS_INTERFACE "org.xchatgnome.interface"

static DBusGProxy *remote_object = NULL;
static gchar *opt_open_url = NULL;
static gchar *opt_command = NULL;
static gchar *opt_print = NULL;
static gchar *opt_channel = NULL;
static gchar *opt_server = NULL;
static gchar *opt_info = NULL;
static gchar *opt_prefs = NULL;

static GOptionEntry entries[] = {
  {"url",     'u', 0, G_OPTION_ARG_STRING, &opt_open_url, N_("Open an irc:// url"), N_("irc://server:port/channel")},
  {"command", 'c', 0, G_OPTION_ARG_STRING, &opt_command, N_("Execute a xchat command"), N_("\"Command to execute\"")},
  {"print",   'p', 0, G_OPTION_ARG_STRING, &opt_print,    N_("Prints some text to the current tab/window"), N_("\"Text to print\"")},
  {"channel", 'h', 0, G_OPTION_ARG_STRING, &opt_channel,  N_("Change the context to the channel"), N_("channel")},
  {"server",  's', 0, G_OPTION_ARG_STRING, &opt_server,   N_("Change the context to the server"), N_("server")},
  {"info",    'i', 0, G_OPTION_ARG_STRING, &opt_info,     N_("Get some informations from xchat"), N_("id")},
  {"prefs",   'r', 0, G_OPTION_ARG_STRING, &opt_prefs,    N_("Get settings from xchat"), N_("name")},
  { NULL }
};

static void
write_error (gchar *message, GError *err)
{
  if (!err)
    return;
  g_printerr ("%s : %s\n", message, err->message);
  g_error_free (err);
}

int
main (int argc, char **argv)
{
  DBusGConnection *bus = NULL;
  GError *error = NULL;
  GOptionContext *context = NULL;
  
#ifdef ENABLE_NLS
  bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);
#endif

  context = g_option_context_new (NULL);
  g_option_context_add_main_entries (context, entries, GETTEXT_PACKAGE);
  g_option_context_parse (context, &argc, &argv, &error);

  if (error)
  {
    g_printerr (_("xchat-gnome-remote: %s\n"
                  "Try `xchat-gnome-remote --help' for more information\n"),
                error->message);
    return 1;
  }

  g_type_init ();

  bus = dbus_g_bus_get (DBUS_BUS_SESSION, &error);
  if (!bus)
  {
    write_error (_("Couldn't connect to session bus"), error);
    return 1;
  }
  
  remote_object = dbus_g_proxy_new_for_name (bus,
                                             DBUS_SERVICE,
                                             DBUS_OBJECT,
                                             DBUS_INTERFACE);

  if (opt_open_url)
  {
    gchar *command;
    command = g_strdup_printf ("url %s", opt_open_url);
    if (!dbus_g_proxy_call (remote_object, "command", &error,
                            G_TYPE_STRING, command, G_TYPE_INVALID,
                            G_TYPE_INVALID))
    {
      if (error->domain == DBUS_GERROR && error->code == DBUS_GERROR_SERVICE_UNKNOWN)
      {
        /* No xchat running... */
        if (fork() == 0)
        {
          gchar *url = g_strdup_printf ("--url=%s", opt_open_url);
          execl (PREFIX"/bin/"PACKAGE_NAME, PACKAGE_NAME, "-a", url, NULL);
        }
      } else
        write_error (_("Failed to complete command"), error);
    }
    g_free (command);
  }
  if (opt_channel || opt_server)
    if (!dbus_g_proxy_call (remote_object, "SetContext", &error,
                            G_TYPE_STRING, opt_server,
                            G_TYPE_STRING, opt_channel, G_TYPE_INVALID,
                            G_TYPE_INVALID))
      write_error (_("Failed to complete SetContext"), error);
  if (opt_command)
    if (!dbus_g_proxy_call (remote_object, "command", &error,
                            G_TYPE_STRING, opt_command, G_TYPE_INVALID,
                            G_TYPE_INVALID))
      write_error (_("Failed to complete command"), error);
  if (opt_print)
    if (!dbus_g_proxy_call (remote_object, "print", &error,
                            G_TYPE_STRING, opt_print, G_TYPE_INVALID,
                            G_TYPE_INVALID))
      write_error (_("Failed to complete print"), error);
  if (opt_info)
  {
    gchar *info;
    if (!dbus_g_proxy_call (remote_object, "GetInfo", &error,
                            G_TYPE_STRING, opt_info, G_TYPE_INVALID,
                            G_TYPE_STRING, &info, G_TYPE_INVALID))
      write_error (_("Failed to complete GetInfo"), error);
    else
      g_print ("%s = %s\n", opt_info, info);
  }
  if (opt_prefs)
  {
    gchar *str;
    int type, i;
    if (!dbus_g_proxy_call (remote_object, "GetPrefs", &error,
                            G_TYPE_STRING, opt_prefs, G_TYPE_INVALID,
                            G_TYPE_INT, &type,
                            G_TYPE_STRING, &str,
                            G_TYPE_INT, &i, G_TYPE_INVALID))
      write_error (_("Failed to complete GetPrefs"), error);
    else
    {
      if (type == 0)
        g_print (_("%s doesn't exist\n"), opt_prefs);
      else if (type == 1)
        g_print ("%s = %s\n", opt_prefs, str);
      else if (type == 2)
        g_print ("%s = %d\n", opt_prefs, i);
      else
        g_print ("%s = %s\n", opt_prefs, i ? "TRUE" : "FALSE");
    }
  }
  
  g_object_unref (G_OBJECT (remote_object));
  return 0;
}
